config/critters-webpack-plugin.js

```diff
@@ -0,0 +1,304 @@
+      compilation.hooks.htmlWebpackPluginAfterHtmlProcessing.tapAsync(PLUGIN_NAME, (htmlPluginData, callback) => {
+        // Parse the generated HTML in a DOM we can mutate
+        const document = parse5.parse(htmlPluginData.html, PARSE5_OPTS);
+        makeDomInteractive(document);
+
+        let externalStylesProcessed = Promise.resolve();
+
+        // `external:false` skips processing of external sheets
+        if (this.options.external!==false) {
+          const externalSheets = document.querySelectorAll('link[rel="stylesheet"]');
+          externalStylesProcessed = Promise.all(externalSheets.map(
+            link => this.embedLinkedStylesheet(link, compilation, outputPath)
+          ));
+        }
+
+        externalStylesProcessed
+          .then(() => {
+            // go through all the style tags in the document and reduce them to only critical CSS
+            const styles = document.querySelectorAll('style');
+            return Promise.all(styles.map(style => this.processStyle(style, document)));
+          })
+          .then(() => {
+            // serialize the document back to HTML and we're done
+            const html = parse5.serialize(document, PARSE5_OPTS);
+            callback(null, { html });
+          })
+          .catch(callback);
+      });
+    });
+  }
+
+  /** Inline the target stylesheet referred to by a <link rel="stylesheet"> (assuming it passes `options.filter`) */
+  embedLinkedStylesheet(link, compilation, outputPath) {
+    const href = link.getAttribute('href');
+    const document = link.ownerDocument;
+
+    // skip filtered resources, or network resources if no filter is provided
+    if (this.urlFilter ? this.urlFilter(href) : href.match(/^(https?:)?\/\//)) return Promise.resolve();
+
+    // path on disk
+    const filename = path.resolve(outputPath, href.replace(/^\//, ''));
+
+    // try to find a matching asset by filename in webpack's output (not yet written to disk)
+    const asset = compilation.assets[path.relative(outputPath, filename).replace(/^\.\//, '')];
+
+    // wait for a disk read if we had to go to disk
+    const promise = asset ? Promise.resolve(asset.source()) : readFile(filename, 'utf8');
+    return promise.then(sheet => {
+    // the reduced critical CSS gets injected into a new <style> tag
+      const style = document.createElement('style');
+      style.appendChild(document.createTextNode(sheet));
+      link.parentNode.insertBefore(style, link.nextSibling);
+
+      // drop a reference to the original URL onto the tag (used for reporting to console later)
+      style.$$name = href;
+
+      // the `async` option changes any critical'd <link rel="stylesheet"> tags to async-loaded equivalents
+      if (this.options.async) {
+        link.setAttribute('rel', 'preload');
+        link.setAttribute('as', 'style');
+        link.setAttribute('onload', "this.rel='stylesheet'");
```

Comment:
What does this `onload` handler do o_O

Reply:
swaps the preload to a stylesheet once it's loaded:
https://www.filamentgroup.com/lab/async-css.html